### PR TITLE
cli(ticdc): support multiple TiCDC clusters to replicate data from one TiDB

### PR DESCRIPTION
**Pull request by sdojjy**
### What problem does this PR solve?

Issue Number: ref #5663

### What is changed and how it works?
1. support multiple TiCDC clusters to replicate data from one TiDB
2.  CDC cli use open api to communicate with cdc server 
3. Add  open API v2 to support more changefeed parameters

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
https://pingcap.feishu.cn/wiki/wikcn9kaQukMcCCS3KNvlr2HXTg?useNewLarklet=1

### Release note 

```release-note
cdc cli use open API to communicate with server,  and the ticdc metadata stored in ETCD will be auto migrated to the new schema version.
```

### CODE REVIEW

File: cdc/api/v2/unsafe.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// ResolveLock resolve locks in regions
+func (h *OpenAPIV2) ResolveLock(c *gin.Context) {
+	var resolveLockReq ResolveLockReq
+	if err := c.BindJSON(&resolveLockReq); err != nil {
+		_ = c.Error(cerror.ErrAPIInvalidParam.Wrap(err))
+		return
+	}
+	up := h.capture.UpstreamManager.GetDefaultUpstream()
```
Comment (sdojjy) [inline cdc/api/v2/unsafe.go:52]: support multi upstream

File: cdc/api/v2/unsafe.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.Status(http.StatusNoContent)
+}
+
+// DeleteServiceGcSafePoint Delete CDC service GC safepoint in PD
+func (h *OpenAPIV2) DeleteServiceGcSafePoint(c *gin.Context) {
+	up := h.capture.UpstreamManager.GetDefaultUpstream()
```
Comment (sdojjy) [inline cdc/api/v2/unsafe.go:66]: multi upstream

File: cdc/capture/capture.go [hunk 3]
```diff
@@ -391,16 +413,24 @@ func (c *Capture) campaignOwner(ctx cdcContext.Context) error {
...
 			c.MessageRouter.AddPeer(captureID, addr)
 		})
 		globalState.SetOnCaptureRemoved(func(captureID model.CaptureID) {
 			c.MessageRouter.RemovePeer(captureID)
 		})
+		globalState.SetOnUpstreamAdded(func(upstreamID model.UpstreamID, info *model.UpstreamInfo) {
+			c.UpstreamManager.AddUpstream(upstreamID, info)
+		})
+		globalState.SetOnUpstreamRemoved(func(upstreamID model.UpstreamID) {
+			c.UpstreamManager.RemoveUpstream(upstreamID)
```
Comment (sdojjy) [inline cdc/capture/capture.go:428]: upstream will be removed twice, owner and processor

File: cdc/processor/processor.go [hunk 4]
```diff
@@ -315,8 +315,15 @@ func isProcessorIgnorableError(err error) bool {
 // the `state` parameter is sent by the etcd worker, the `state` must be a snapshot of KVs in etcd
 // The main logic of processor is in this function, including the calculation of many kinds of ts, maintain table pipeline, error handling, etc.
 func (p *processor) Tick(ctx cdcContext.Context, state *orchestrator.ChangefeedReactorState) (orchestrator.ReactorState, error) {
+	// skip this tick
+	if err := p.upstream.Error(); err != nil {
```
Comment (sdojjy) [inline cdc/processor/processor.go:319]: how to re-create a new upstream?

File: pkg/api/v2/tso.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+// newTso returns tso
+func newTso(c *APIV2Client) *tso {
+	return &tso{
+		client: c.RESTClient(),
+	}
+}
+
+// Get returns the pd tso
+func (c *tso) Get(ctx context.Context) (*v2.Tso, error) {
```
Comment (sdojjy) [inline pkg/api/v2/tso.go:46]: tso support multi upstream

File: pkg/cmd/cli/cli_changefeed_create.go [hunk 6]
```diff
@@ -282,151 +241,103 @@ func (o *createChangefeedOptions) validate(ctx context.Context, cmd *cobra.Comma
 	}
 
 	switch o.commonChangefeedOptions.sortEngine {
-	case model.SortUnified, model.SortInMemory:
-	case model.SortInFile:
-		// obsolete. But we keep silent here. We create a Unified Sorter when the owner/processor sees this option
-		// for backward-compatibility.
+	case model.SortUnified:
```
Comment (sdojjy) [inline pkg/cmd/cli/cli_changefeed_create.go:244]: why remove memory engine?

File: pkg/cmd/cli/cli_changefeed_resume.go [hunk 7]
```diff
@@ -53,49 +48,43 @@ func (o *resumeChangefeedOptions) addFlags(cmd *cobra.Command) {
...
+		for _, cf := range *cfs {
+			if cf.ID == o.changefeedID {
+				checkpointTs = cf.CheckpointTSO
+			}
+		}
+		if checkpointTs == 0 {
+			return errors.ErrChangeFeedNotExists
+		}
+
+		tso, err := o.apiV2Client.Tso().Get(ctx)
```
Comment (sdojjy) [inline pkg/cmd/cli/cli_changefeed_resume.go:82]: support multi upstream

**Review by lonng — COMMENTED** [review]
This PR is too huge, please split it into small PRs.

**Review by nongfushanquan — COMMENTED** [review]
Please add all  test results

**Review by nongfushanquan — APPROVED** [review]

**Review by asddongmen — APPROVED** [review]

File: pkg/etcd/client_for_api.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import (
+	"context"
+
+	"github.com/pingcap/tiflow/cdc/model"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+)
+
+// CDCEtcdClientForAPI extracts CDCEtcdClients's method used for apiv2
+// TODO: refactor CDCEtcdClient as interface, and use it for v1
+type CDCEtcdClientForAPI interface {
```
Comment (overvenus) [inline pkg/etcd/client_for_api.go:25]: Why not just name it `CDCEtcdClient`?

File: cdc/capture/capture.go [hunk 9]
```diff
@@ -102,30 +105,41 @@ type captureImpl struct {
...
 		grpcService:         grpcService,
 		cancel:              func() {},
-		pdEndpoints:         pdEnpoints,
+		pdEndpoints:         pdEndpoints,
 		newProcessorManager: processor.NewManager,
 		newOwner:            owner.NewOwner,
+
+		migrator: migrate.NewMigrator(
+			etcdClient,
+			pdEndpoints, conf),
```
Comment (overvenus) [inline cdc/capture/capture.go:131]:
```suggestion
		migrator: migrate.NewMigrator(etcdClient, pdEndpoints, conf),
```

File: cdc/capture/capture.go [hunk 10]
```diff
@@ -732,3 +754,8 @@ func (c *captureImpl) StatusProvider() owner.StatusProvider {
 	}
 	return owner.NewStatusProvider(c.owner)
 }
+
+// IsReady returns if the cdc server is ready
```
Comment (overvenus) [inline cdc/capture/capture.go:758]: Comments should also be added to the interface.

File: scripts/check-diff-line-width.sh [hunk 11]
```diff
@@ -26,6 +26,7 @@ ERROR_THRESHOLD=100
 git --no-pager diff $BASE_HASH -U0 -- cdc pkg cmd \
 	-- ':(exclude)*_gen.go' \
 	-- ':(exclude)*_gen_test.go' \
+	-- ':(exclude)*_mock.go' \
 	-- ':(exclude)*.pb.go' |
 	-- ':(exclude)*_mock.pb.go' |
```
Comment (overvenus) [inline scripts/check-diff-line-width.sh:31]:
```suggestion
```

It looks like a mistake, there is no such in code base, please help remove this line.

File: pkg/etcd/client_for_api.go (hunk 8 above)
Comment (sdojjy) [reply pkg/etcd/client_for_api.go:25]: here we only mock a part of `CDCEtcdClient` functions for API, later we will add all method an then rename this interface.

File: cdc/capture/capture.go (hunk 10 above)
Comment (sdojjy) [reply cdc/capture/capture.go:756]: done

**Pull request merged**